packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -71,46 +60,223 @@ class Intent with Diagnosticable {
 ///    up key combinations in order to invoke actions.
 ///  * [Actions], which is a widget that defines a map of [Intent] to [Action]
 ///    and allows redefining of actions for its descendants.
-///  * [ActionDispatcher], a class that takes an [Action] and invokes it using a
-///    [FocusNode] for context.
-abstract class Action with Diagnosticable {
-  /// A const constructor for an [Action].
-  ///
-  /// The [intentKey] parameter must not be null.
-  const Action(this.intentKey) : assert(intentKey != null);
+///  * [ActionDispatcher], a class that takes an [Action] and invokes it, passing
+///    a given [Intent].
+abstract class Action<T extends Intent> with Diagnosticable {
+  final ObserverList<ActionListenerCallback> _listeners = ObserverList<ActionListenerCallback>();
+
+  /// Gets the type of intent this action responds to.
+  Type get intentType => T;
 
-  /// The unique key for this action.
+  /// Returns true if the action is enabled and is ready to be invoked.
   ///
-  /// This key will be used to map to this action in an [ActionDispatcher].
-  final LocalKey intentKey;
+  /// This will be called by the [ActionDispatcher] before attempting to invoke
+  /// the action.
+  ///
+  /// If the enabled state changes, overriding subclasses must call
+  /// [notifyActionListeners] to notify any listeners of the change.
+  bool get enabled => true;
 
   /// Called when the action is to be performed.
   ///
-  /// This is called by the [ActionDispatcher] when an action is accepted by a
-  /// [FocusNode] by returning true from its `onAction` callback, or when an
-  /// action is invoked using [ActionDispatcher.invokeAction].
+  /// This is called by the [ActionDispatcher] when an action is invoked via
+  /// [Actions.invoke], or when an action is invoked using
+  /// [ActionDispatcher.invokeAction] directly.
   ///
   /// This method is only meant to be invoked by an [ActionDispatcher], or by
-  /// subclasses.
+  /// its subclasses, and only when [enabled] is true.
+  @protected
+  Object invoke(covariant T intent);
```

Comment:
nit: Can the docs above say something about the return type: If I implement my own action, what should I return here?

Reply:
Yes, I added a comment, and a small code sample.